### PR TITLE
Fine grain dof adjustment

### LINKED ISSUE #285: Queen's Feature Request - Finer Controls With Arrow Keys
When working with manual tracking, the ability to navigate through the image using arrow keys would be highly beneficial. This functionality currently requires manually dragging the arrows. By enabling arrow key (or any key up to your discretion) navigation, users can seamlessly move the radiograph and customize the value of each key press. This approach improves efficiency by allowing quick adjustments and precise control. Add option to set the increment of the rotation and translation for each arrow press.

### PR DESCRIPTION
**Pull request by NicerNewerCar**
Adds support for manipulating DRR position using the arrow keys.
Controls:
* `Left`  Select previous DOF. 
* `Right` Select next DOF. 
* `Up`  Increment selected DOF. 
* `Down`  Decrement selected DOF. 

Selected DOF is bolded on the timeline

Vary the step size by going to `Tracking-> Advanced Settings -> Step Size` defaults to `1mm/deg`
Closes #285 
arrow_input

### CODE REVIEW

**Comment by amymmorton** [comment]
cool - lgtm
should @mjrainbow test too?

**Review by amymmorton — APPROVED** [review]

File: autoscoper/src/ui/AutoscoperMainWindow.h [hunk 1]
```diff
@@ -220,6 +222,10 @@ class AutoscoperMainWindow : public QMainWindow
   std::vector<unsigned int> textures;
   void reset_graph();
 
+  // Arrow key stepping
+  unsigned int stepDOF = 0;
```
Comment (jcfr) [inline autoscoper/src/ui/AutoscoperMainWindow.h:226]: * `unsigned int` -> `int`
* rename from `stepDOF` to `currentDOF`
* move into `TimelineDockWidget`

File: autoscoper/src/ui/TimelineDockWidget.cpp [hunk 2]
```diff
@@ -292,6 +292,46 @@ void TimelineDockWidget::on_spinBox_LastFrame_valueChanged(int d)
   }
 }
 
+void TimelineDockWidget::updateBolding(const unsigned int dof, const bool state)
+{
+  QFont* font;
```
Comment (jcfr) [inline autoscoper/src/ui/TimelineDockWidget.cpp:297]:
```suggestion
  QFont font;
```

and simplify code below accordingly.

File: autoscoper/src/ui/TimelineDockWidget.cpp [hunk 3]
```diff
@@ -292,6 +292,46 @@ void TimelineDockWidget::on_spinBox_LastFrame_valueChanged(int d)
...
+  switch (dof) {
+    case 0:
+      font = new QFont(dock->checkBox_X->font());
+      font->setBold(state);
+      dock->checkBox_X->setFont(*font);
```
Comment (jcfr) [inline autoscoper/src/ui/TimelineDockWidget.cpp:302]:
```suggestion
      font = dock->checkBox_X->font();
      font->setBold(state);
      dock->checkBox_X->setFont(font);
```

**Review by DajungYoon — APPROVED** [review]

File: autoscoper/src/ui/TimelineDockWidget.h [hunk 4]
```diff
@@ -100,6 +100,10 @@ class TimelineDockWidget : public QDockWidget
 
   std::vector<std::pair<IKeyCurve*, IKeyCurve::iterator>>* getCopiedNodes() { return &copied_nodes; }
 
+  void updateBolding(const unsigned int dof, const bool state);
+  void removeBolding(const unsigned int dof) { updateBolding(dof, false); }
+  void addBolding(const unsigned int dof) { updateBolding(dof, true); }
```
Comment (jcfr) [inline autoscoper/src/ui/TimelineDockWidget.h:105]: Simplify API by making `updateBolding` private and having only one function:

```
void setCurrentDOF(int dof)
{
  if (dof < 0) {
    dof = 0;
  }
  if (dof >= NUMBER_OF_DEGREES_OF_FREEDOM) {
    dof = NUMBER_OF_DEGREES_OF_FREEDOM - 1;
  }
  if (dof == this->currentDOF) {
    return;
  }
  this->updateBolding(this->currentDOF, false);
  this->updateBolding(dof, true);
  this->currentDOF = dof;
}
```

File: autoscoper/src/ui/AutoscoperMainWindow.cpp [hunk 5]
```diff
@@ -2399,6 +2407,43 @@ void AutoscoperMainWindow::key_minus_pressed()
...
+void AutoscoperMainWindow::key_left_pressed()
+{
+  // Move to the prev DOF
+  timeline_widget->removeBolding(stepDOF);
+  if (stepDOF == 0) {
+    stepDOF = 5;
+  } else {
+    stepDOF--;
+  }
+  timeline_widget->addBolding(stepDOF);
```
Comment (jcfr) [inline autoscoper/src/ui/AutoscoperMainWindow.cpp:2419]:
```suggestion
  int updatedDOF = this->timeline_widget->currentDOF();
  updatedDOF--;
  if (updatedDOF < 0) {
    updatedDOF = TimelineDockWidget::NUMBER_OF_DEGREES_OF_FREEDOM - 1;
  }
  this->timeline_widget->setCurrentDOF(updatedDOF);
```

File: autoscoper/src/ui/AutoscoperMainWindow.cpp [hunk 6]
```diff
@@ -2399,6 +2407,43 @@ void AutoscoperMainWindow::key_minus_pressed()
...
+
+void AutoscoperMainWindow::key_right_pressed()
+{
+  // Move to the next DOF
+  timeline_widget->removeBolding(stepDOF);
+  stepDOF++;
+  if (stepDOF >= 6) {
+    stepDOF = 0;
+  }
+  timeline_widget->addBolding(stepDOF);
```
Comment (jcfr) [inline autoscoper/src/ui/AutoscoperMainWindow.cpp:2430]:
```suggestion
  int updatedDOF = timeline_widget->currentDOF();
  updatedDOF++;
  if (updatedDOF >= TimelineDockWidget::NUMBER_OF_DEGREES_OF_FREEDOM) {
    updatedDOF = 0;
  }
  timeline_widget->setCurrentDOF(updatedDOF);
```

File: autoscoper/src/ui/AutoscoperMainWindow.cpp [hunk 7]
```diff
@@ -2399,6 +2407,43 @@ void AutoscoperMainWindow::key_minus_pressed()
...
+}
+
+void AutoscoperMainWindow::key_up_pressed()
+{
+  std::vector<double> curPose = getPose(-1, -1);
+  curPose[stepDOF] = curPose[stepDOF] + stepVal;
```
Comment (jcfr) [inline autoscoper/src/ui/AutoscoperMainWindow.cpp:2436]:
```suggestion
  std::vector<double> curPose = getPose(-1, -1);
  int currentDOF = this->timeline_widget->currentDOF();
  curPose[currentDOF] = curPose[currentDOF] + stepVal;
```

**Review by jcfr — COMMENTED** [review]
Beside of few nitpicks, I suggest to simplify the logic by doing the bookkeeping of the DOF being updated directly int the timeline widget.

File: autoscoper/src/ui/AutoscoperMainWindow.cpp [hunk 8]
```diff
@@ -833,23 +833,31 @@ void AutoscoperMainWindow::loadFilterSettings(int camera, QString filename)
...
+
   std::vector<double> pose(6, 0);
   pose[0] = (*tracker->trial()->getXCurve(volume))(frame);
   pose[1] = (*tracker->trial()->getYCurve(volume))(frame);
   pose[2] = (*tracker->trial()->getZCurve(volume))(frame);
   Quatf q = (*tracker->trial()->getQuatCurve(volume))(frame);
-  pose[3] = q.z;
-  pose[4] = q.y;
-  pose[5] = q.x;
+  Vec3f euler = q.toEuler();
```
[the remark this answers is not in the document]
Comment (NicerNewerCar) [reply autoscoper/src/ui/AutoscoperMainWindow.cpp]: That seems to be the only missing conversion 
See https://github.com/BrownBiomechanics/Autoscoper/pull/296

**Pull request merged**